### PR TITLE
Add half precision floating point support to `StreamPeer` and `FileAccess`

### PR DESCRIPTION
**Pull request by pafuent**
Closes godotengine/godot-proposals#5983

Adds put/get methods to `StreamPeer` that handles half precision floating point values.
Adds endode/decode half precision floating point to `marshalls`.

This PR is rebase into master of #78872. Also I added unit tests to it.

### CODE REVIEW

**Review by radiantgurl — COMMENTED** [review]
Shouldnt we extend it to FileAccess and PackedByteArray now given the chance to do so?

**Comment by pafuent** [comment]
`PackedByteArray` already have those:
`encode_half` binded [here](https://github.com/godotengine/godot/blob/master/core/variant/variant_call.cpp#L2431) and implemented [here](https://github.com/godotengine/godot/blob/master/core/variant/variant_call.cpp#L1002)
`decode_half` binded [here](https://github.com/godotengine/godot/blob/master/core/variant/variant_call.cpp#L2411) and implemented [here](https://github.com/godotengine/godot/blob/master/core/variant/variant_call.cpp#L840)

**Review by radiantgurl — APPROVED** [review]

**Review by radiantgurl — APPROVED** [review]

**Comment by RadiantUwU** [comment]
Topic labels changed, we should change the title and the topics since now it also modifies FileAccess (core)

**Comment by pafuent** [comment]
Friendly remainder

**Review by radiantgurl — APPROVED** [review]

**Review by Calinou — APPROVED** [review]
Tested locally, it works as expected.

Thanks for adding unit tests :slightly_smiling_face: 

```gdscript
extends Node2D


func _ready() -> void:
	var fa := FileAccess.open("user://output.bin", FileAccess.WRITE_READ)
	fa.store_half(0.123456)
	fa.store_float(0.123456)
	fa.store_double(0.123456)

	fa.store_double(-12345)
	fa.store_float(-12345)
	fa.store_half(-12345)

	OS.shell_open(ProjectSettings.globalize_path("user://"))
	get_tree().quit()
```

Using [ImHex](https://imhex.werwolv.net/) to view the file written by Godot:

image

image

Note the lack of precision on the second example, as half floats can't represent all integer values outside of `[-2048; 2048]`.

File: tests/core/io/test_file_access.h [hunk 1]
```diff
@@ -107,6 +107,98 @@ TEST_CASE("[FileAccess] Get as UTF-8 String") {
...
+	// BigEndian Hex: 0x40490E56
+	// LittleEndian Hex: 0x560E4940
+	float value = 3.1415f;
+
+	SUBCASE("Little Endian") {
+		const String file_path = TestUtils::get_data_path("floating_point_little_endian.bin");
+		const String file_path_new = TestUtils::get_data_path("floating_point_little_endian_new.bin");
+
+		Ref<FileAccess> f = FileAccess::open(file_path, FileAccess::READ);
+		REQUIRE(!f.is_null());
```
Comment (AThousandShips) [inline tests/core/io/test_file_access.h:121]:
```suggestion
		REQUIRE(f.is_valid());
```

File: tests/core/io/test_file_access.h [hunk 2]
```diff
@@ -107,6 +107,98 @@ TEST_CASE("[FileAccess] Get as UTF-8 String") {
...
+		CHECK_EQ(f->get_float(), value);
+
+		Ref<FileAccess> fw = FileAccess::open(file_path_new, FileAccess::WRITE);
+		REQUIRE(!fw.is_null());
```
Comment (AThousandShips) [inline tests/core/io/test_file_access.h:125]:
```suggestion
		REQUIRE(fw.is_valid());
```

File: tests/core/io/test_file_access.h [hunk 3]
```diff
@@ -107,6 +107,98 @@ TEST_CASE("[FileAccess] Get as UTF-8 String") {
...
+
+		DirAccess::remove_file_or_error(file_path_new);
+	}
+
+	SUBCASE("Big Endian") {
+		const String file_path = TestUtils::get_data_path("floating_point_big_endian.bin");
+		const String file_path_new = TestUtils::get_data_path("floating_point_big_endian_new.bin");
+
+		Ref<FileAccess> f = FileAccess::open(file_path, FileAccess::READ);
+		REQUIRE(!f.is_null());
```
Comment (AThousandShips) [inline tests/core/io/test_file_access.h:139]:
```suggestion
		REQUIRE(f.is_valid());
```

**Review by AThousandShips — APPROVED** [review]

**Comment by pafuent** [comment]
Friendly reminder

**Review by Faless — APPROVED** [review]
LGTM! Nice work! :rocket:

**Comment by akien-mga** [comment]
It's now queued for merging, thanks for your patience :)

**Comment by Repiteo** [comment]
Thanks!

**Pull request merged**